### PR TITLE
Add AssertOrFailFast

### PR DESCRIPTION
**Pull request by jaredpar**
This introduces `AssertOrFailFast`. This is like `Debug.Assert` but it
will provide a crash dump when the `Assert` fails in CI runs.

### CODE REVIEW

**Comment by jaredpar** [comment]
@dotnet/roslyn-compiler PTAL

File: src/Compilers/Core/Portable/InternalUtilities/Debug.cs [hunk 1]
```diff
@@ -23,5 +24,39 @@ public static void AssertNotNull<T>([NotNull] T value)
...
+        /// only invariants in our code. When it triggers that providse a nice stack trace for 
+        /// investigation. Generally that is enough.
+        /// 
+        /// There are cases for which a stack is not enough and we need a full heap dump to 
+        /// investigate the failure. This method takes care of that. The behavior is that when running
+        /// in our CI environment if the assert triggers we will rudely crash the process and 
+        /// produce a heap dump for investigation.
+        /// </summary>
+        [Conditional("DEBUG")]
+        internal static void AssertOrFailFast(bool condition, string? message = null)
```
Comment (sharwell) [inline src/Compilers/Core/Portable/InternalUtilities/Debug.cs:39]:
```suggestion
        internal static void AssertOrFailFast([DoesNotReturnIf(false)] bool condition, string? message = null)
```

File: src/Compilers/Core/Portable/InternalUtilities/Debug.cs [hunk 2]
```diff
@@ -23,5 +24,39 @@ public static void AssertNotNull<T>([NotNull] T value)
...
+
+        /// <summary>
+        /// Generally <see cref="Debug.Assert(bool)"/> is a sufficient method for enforcing DEBUG 
+        /// only invariants in our code. When it triggers that providse a nice stack trace for 
+        /// investigation. Generally that is enough.
+        /// 
+        /// There are cases for which a stack is not enough and we need a full heap dump to 
+        /// investigate the failure. This method takes care of that. The behavior is that when running
+        /// in our CI environment if the assert triggers we will rudely crash the process and 
+        /// produce a heap dump for investigation.
```
Comment (sharwell) [inline src/Compilers/Core/Portable/InternalUtilities/Debug.cs:36]:
```suggestion
        /// <para>There are cases for which a stack is not enough and we need a full heap dump to 
        /// investigate the failure. This method takes care of that. The behavior is that when running
        /// in our CI environment if the assert triggers we will rudely crash the process and 
        /// produce a heap dump for investigation.</para>
```

**Review by sharwell — APPROVED** [review]

File: src/Compilers/Core/Portable/InternalUtilities/Debug.cs [hunk 3]
```diff
@@ -23,5 +24,43 @@ public static void AssertNotNull<T>([NotNull] T value)
...
+        /// <para>There are cases for which a stack is not enough and we need a full heap dump to 
+        /// investigate the failure. This method takes care of that. The behavior is that when running
+        /// in our CI environment if the assert triggers we will rudely crash the process and 
+        /// produce a heap dump for investigation.</para>
+        /// </summary>
+        [Conditional("DEBUG")]
+        internal static void AssertOrFailFast([DoesNotReturnIf(false)] bool condition, string? message = null)
+        {
+#if NET20
+            Debug.Assert(false);
```
Comment (sharwell) [inline src/Compilers/Core/Portable/InternalUtilities/Debug.cs:42]:
```suggestion
            Debug.Assert(condition);
```

File: src/Compilers/Core/Portable/InternalUtilities/Debug.cs [hunk 4]
```diff
@@ -23,5 +24,43 @@ public static void AssertNotNull<T>([NotNull] T value)
...
+                    Console.WriteLine(stackTrace);
+
+                    // Use FailFast so that the process fails rudely and goes through 
+                    // windows error reporting (on Windows at least). This will allow our 
+                    // Helix environment to capture crash dumps for future investigation
+                    Environment.FailFast(message);
+                }
+                else
+                {
+                    Debug.Assert(false);
```
Comment (sharwell) [inline src/Compilers/Core/Portable/InternalUtilities/Debug.cs:60]:
```suggestion
                    Debug.Assert(false, message);
```

**Comment by jaredpar** [comment]
@AlekseyTs this should start giving us dumps for that assert failure

File: src/Compilers/Core/Portable/InternalUtilities/Debug.cs [hunk 5]
```diff
@@ -23,5 +24,43 @@ public static void AssertNotNull<T>([NotNull] T value)
...
+        /// </summary>
+        [Conditional("DEBUG")]
+        internal static void AssertOrFailFast([DoesNotReturnIf(false)] bool condition, string? message = null)
+        {
+#if NET20 || NETSTANDARD1_3
+            Debug.Assert(condition);
+#else
+            if (!condition)
+            {
+                if (!string.IsNullOrEmpty(Environment.GetEnvironmentVariable("HELIX_DUMP_FOLDER")))
```
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/InternalUtilities/Debug.cs:46]: Is this our mechanism for deciding that we're running in a CI environment? I am not sure if this variable is set during the build or during "single-machine" tests.

**Review by RikkiGibson — APPROVED** [review]

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 6)

**Pull request merged**